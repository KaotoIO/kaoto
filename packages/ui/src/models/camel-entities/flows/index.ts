export * from './route';
export * from './step';
