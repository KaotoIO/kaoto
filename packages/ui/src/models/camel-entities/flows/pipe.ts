import { Pipe as PipeModel } from '@kaoto-next/camel-catalog/types';
import { v4 as uuidv4 } from 'uuid';
import { BaseVisualCamelEntity, EntityType } from '../base-entity';
import { KameletBinding } from './kamelet-binding';

export class Pipe extends KameletBinding implements BaseVisualCamelEntity {
  readonly id = uuidv4();
  type = EntityType.Pipe;

  constructor(public route: Partial<PipeModel> = {}) {
    super();
  }
}
