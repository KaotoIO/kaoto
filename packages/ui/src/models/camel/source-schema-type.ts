export enum SourceSchemaType {
  Route = 'Route',
  Integration = 'Integration',
  Kamelet = 'Kamelet',
  KameletBinding = 'KameletBinding',
  Pipe = 'Pipe',
}
