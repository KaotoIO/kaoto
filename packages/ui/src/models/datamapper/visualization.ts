import { IDocument, IField, PrimitiveDocument } from './document';
import { ExpressionItem, FieldItem, IFunctionDefinition, MappingItem, MappingParentType, MappingTree } from './mapping';
import { DocumentType, NodePath } from './path';
import { RefObject } from 'react';

export interface NodeData {
  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export interface TargetNodeData extends NodeData {
  mappingTree: MappingTree;
  mapping?: MappingParentType;
}

export type SourceNodeDataType = DocumentNodeData | FieldNodeData;
export type TargetNodeDataType = TargetDocumentNodeData | TargetFieldNodeData;

export class DocumentNodeData implements NodeData {
  constructor(document: IDocument) {
    this.title = document.documentId;
    this.id = `doc-${document.documentType}-${document.documentId}`;
    this.path = NodePath.fromDocument(document.documentType, document.documentId);
    this.document = document;
    this.isSource = document.documentType !== DocumentType.TARGET_BODY;
    this.isPrimitive = document instanceof PrimitiveDocument;
  }

  document: IDocument;
  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export class TargetDocumentNodeData extends DocumentNodeData implements TargetNodeData {
  constructor(document: IDocument, mappingTree: MappingTree) {
    super(document);
    this.mappingTree = mappingTree;
    this.mapping = mappingTree;
  }
  mappingTree: MappingTree;
  mapping: MappingTree;
}

export class FieldNodeData implements NodeData {
  constructor(
    public parent: NodeData,
    public field: IField,
  ) {
    this.title = field.name;
    this.id = field.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.isSource = parent.isSource;
    this.isPrimitive = parent.isPrimitive;
  }

  title: string;
  id: string;
  path: NodePath;
  isSource: boolean;
  isPrimitive: boolean;
}

export class TargetFieldNodeData extends FieldNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public field: IField,
    public mapping?: FieldItem,
  ) {
    super(parent, field);
    this.mappingTree = parent.mappingTree;
  }
  mappingTree: MappingTree;
}

export class MappingNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public mapping: MappingItem,
  ) {
    this.title = mapping.name;
    this.id = mapping.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.isPrimitive = parent.isPrimitive;
    this.mappingTree = parent.mappingTree;
  }

  title: string;
  id: string;
  path: NodePath;
  isSource = false;
  isPrimitive: boolean;
  mappingTree: MappingTree;
}

export class FieldItemNodeData extends MappingNodeData {
  constructor(
    public parent: TargetNodeData,
    public mapping: FieldItem,
  ) {
    super(parent, mapping);
    this.title = mapping.field.name;
    this.field = mapping.field;
  }
  public field: IField;
}

export class AddMappingNodeData implements TargetNodeData {
  constructor(
    public parent: TargetNodeData,
    public field: IField,
  ) {
    const ID_PREFIX = 'add-mapping-';
    this.id = ID_PREFIX + field.id;
    this.path = NodePath.childOf(parent.path, this.id);
    this.title = field.name;
    this.isPrimitive = parent.isPrimitive;
    this.mappingTree = parent.mappingTree;
  }
  id: string;
  isPrimitive: boolean;
  isSource = false;
  mappingTree: MappingTree;
  path: NodePath;
  title: string;
}

class SimpleNodePath extends NodePath {
  constructor(public path: string) {
    super();
  }
  toString() {
    return this.path;
  }
}
export class EditorNodeData implements NodeData {
  constructor(public mapping: ExpressionItem) {}
  id: string = 'editor';
  isPrimitive: boolean = false;
  isSource: boolean = false;
  path: NodePath = new SimpleNodePath('Editor');
  title: string = 'Editor';
}

export class FunctionNodeData implements NodeData {
  constructor(public functionDefinition: IFunctionDefinition) {
    this.id = functionDefinition.name;
    this.title = functionDefinition.displayName;
    this.path = new SimpleNodePath('Func:' + functionDefinition.name);
  }

  id: string;
  isPrimitive: boolean = false;
  isSource: boolean = true;
  path: NodePath;
  title: string;
}

export interface IMappingLink {
  sourceNodePath: string;
  targetNodePath: string;
  isSelected: boolean;
}

export interface NodeReference {
  path: string;
  isSource: boolean;
  headerRef: HTMLDivElement | null;
  containerRef: HTMLDivElement | null;
}

export type LineCoord = {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
};

export type LineProps = LineCoord & {
  sourceNodePath: string;
  targetNodePath: string;
  isSelected?: boolean;
  svgRef?: RefObject<SVGSVGElement>;
};
