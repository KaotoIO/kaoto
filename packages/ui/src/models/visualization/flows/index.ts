export * from './camel-catalog.service';
export * from './pipe-visual-entity';
export * from './kamelet-binding-visual-entity';
export * from './kamelet-visual-entity';
export * from './camel-route-visual-entity';
