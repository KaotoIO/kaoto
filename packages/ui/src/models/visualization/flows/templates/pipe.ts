import { getCamelRandomId } from '../../../../camel-utils/camel-random-id';

export const pipeTemplate = () => {
  return `apiVersion: camel.apache.org/v1
kind: Pipe
metadata:
  name: ${getCamelRandomId('pipe')}
spec:
  source:
    properties:
      message: hello
    ref:
      kind: Kamelet
      apiVersion: camel.apache.org/v1
      name: timer-source
  sink:
    ref:
      kind: Kamelet
      apiVersion: camel.apache.org/v1
      name: log-sink`;
};
