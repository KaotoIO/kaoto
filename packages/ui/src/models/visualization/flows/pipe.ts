import { Pipe as PipeModel } from '@kaoto-next/camel-catalog/types';
import { v4 as uuidv4 } from 'uuid';
import { BaseVisualCamelEntity } from '../base-visual-entity';
import { EntityType } from '../../camel-entities/base-entity';
import { KameletBinding } from './kamelet-binding';

export class Pipe extends KameletBinding implements BaseVisualCamelEntity {
  readonly id = uuidv4();
  type = EntityType.Pipe;

  constructor(public route: Partial<PipeModel> = {}) {
    super();
  }
}
