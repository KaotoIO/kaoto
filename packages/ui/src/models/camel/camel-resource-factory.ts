import { SourceSchemaType } from './source-schema-type';
import { CamelResource } from './camel-resource';
import { XmlCamelResourceSerializer, YamlCamelResourceSerializer } from '../../serializers';
import { CamelRouteResource } from './camel-route-resource';
import { CamelKResourceFactory } from './camel-k-resource-factory';

export class CamelResourceFactory {
  /**
   * Creates a CamelResource based on the given {@link type} and {@link source}. If
   * both are not specified, a default empty {@link CamelRouteResource} is created.
   * If only {@link type} is specified, an empty {@link CamelResource} of the given
   * {@link type} is created.
   * @param type
   * @param source
   */
  static createCamelResource(source?: string, type?: SourceSchemaType): CamelResource {
    if (XmlCamelResourceSerializer.isApplicable(source)) {
      return new CamelRouteResource(source, new XmlCamelResourceSerializer());
    }

    const serializer = new YamlCamelResourceSerializer();
    const resource = CamelKResourceFactory.getCamelKResource(serializer.parse(source ?? ''), type);

    if (resource) return resource;
    return new CamelRouteResource(source, serializer);
  }
}
