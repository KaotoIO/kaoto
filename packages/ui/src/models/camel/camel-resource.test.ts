import { camelRouteYaml } from '../../stubs/camel-route';
import { integrationJson } from '../../stubs/integration';
import { kameletBindingJson } from '../../stubs/kamelet-binding-route';
import { kameletJson } from '../../stubs/kamelet-route';
import { pipeJson } from '../../stubs/pipe';
import { CamelRouteVisualEntity, PipeVisualEntity } from '../visualization/flows';
import { SourceSchemaType } from './source-schema-type';
import { CamelResourceFactory } from './camel-resource-factory';

describe('CamelResourceFactory.createCamelResource', () => {
  it('should create an empty CamelRouteResource if no args is specified', () => {
    const resource = CamelResourceFactory.createCamelResource();
    expect(resource.getType()).toEqual(SourceSchemaType.Route);
    expect(resource.getEntities()).toEqual([]);
    expect(resource.getVisualEntities()).toEqual([]);
  });

  it('should create an empty CamelRouteResource if no args is specified', () => {
    const resource = CamelResourceFactory.createCamelResource(undefined, SourceSchemaType.Route);
    expect(resource.getType()).toEqual(SourceSchemaType.Route);
    expect(resource.getEntities()).toEqual([]);
    expect(resource.getVisualEntities()).toEqual([]);
  });

  it('should create an empty IntegrationResource if no args is specified', () => {
    const resource = CamelResourceFactory.createCamelResource(undefined, SourceSchemaType.Integration);
    expect(resource.getType()).toEqual(SourceSchemaType.Integration);
    expect(resource.getEntities()).toEqual([]);
    expect(resource.getVisualEntities()).toEqual([]);
  });

  it('should create an empty KameletResource if no args is specified', () => {
    const resource = CamelResourceFactory.createCamelResource(undefined, SourceSchemaType.Kamelet);
    expect(resource.getType()).toEqual(SourceSchemaType.Kamelet);
    expect(resource.getEntities()).toEqual([]);
    expect(resource.getVisualEntities()).toMatchSnapshot();
  });

  it('should create an empty CameletBindingResource if no args is specified', () => {
    const resource = CamelResourceFactory.createCamelResource(undefined, SourceSchemaType.KameletBinding);
    expect(resource.getType()).toEqual(SourceSchemaType.KameletBinding);
    expect(resource.getEntities()).toEqual([]);
    expect(resource.getVisualEntities().length).toEqual(1);
  });

  it('should create an empty PipeResource if no args is specified', () => {
    const resource = CamelResourceFactory.createCamelResource(undefined, SourceSchemaType.Pipe);
    expect(resource.getType()).toEqual(SourceSchemaType.Pipe);
    expect(resource.getEntities()).toEqual([]);
    expect(resource.getVisualEntities().length).toEqual(1);
    const vis = resource.getVisualEntities()[0] as PipeVisualEntity;
    expect(vis.pipe.spec?.source).toBeUndefined();
    expect(vis.pipe.spec?.steps).toBeUndefined();
    expect(vis.pipe.spec?.sink).toBeUndefined();
  });

  it('should create a camel route', () => {
    const resource = CamelResourceFactory.createCamelResource(camelRouteYaml);
    expect(resource.getType()).toEqual(SourceSchemaType.Route);
    expect(resource.getVisualEntities().length).toEqual(1);
    const vis = resource.getVisualEntities()[0] as CamelRouteVisualEntity;
    expect(vis.entityDef.route.from?.uri).toBeDefined();
  });

  // TODO
  it.skip('should create an Integration', () => {
    const resource = CamelResourceFactory.createCamelResource(JSON.stringify(integrationJson));
    expect(resource.getType()).toEqual(SourceSchemaType.Integration);
    expect(resource.getVisualEntities().length).toEqual(2);
  });

  it('should create a Kamelet', () => {
    const resource = CamelResourceFactory.createCamelResource(kameletJson);
    expect(resource.getType()).toEqual(SourceSchemaType.Kamelet);
    expect(resource.getVisualEntities().length).toEqual(1);
  });

  it('should create a KameletBindingPipe', () => {
    const resource = CamelResourceFactory.createCamelResource(JSON.stringify(kameletBindingJson));
    expect(resource.getType()).toEqual(SourceSchemaType.KameletBinding);
    expect(resource.getVisualEntities().length).toEqual(1);
    const vis = resource.getVisualEntities()[0] as PipeVisualEntity;
    expect(vis.pipe.spec?.source?.ref?.name).toEqual('webhook-source');
  });

  it('should create a Pipe', () => {
    const resource = CamelResourceFactory.createCamelResource(JSON.stringify(pipeJson));
    expect(resource.getType()).toEqual(SourceSchemaType.Pipe);
    expect(resource.getVisualEntities().length).toEqual(1);
    const vis = resource.getVisualEntities()[0] as PipeVisualEntity;
    expect(vis.pipe.spec?.source?.ref?.name).toEqual('webhook-source');
  });
});
