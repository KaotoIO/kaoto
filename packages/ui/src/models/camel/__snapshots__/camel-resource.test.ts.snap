// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCamelResource should create an empty KameletResource if no args is specified 1`] = `
[
  {
    "route": {
      "from": {
        "id": "from-1234",
        "parameters": {
          "period": "{{period}}",
          "timerName": "user",
        },
        "steps": [
          {
            "to": "https://random-data-api.com/api/v2/users",
          },
          {
            "to": "kamelet:sink",
          },
        ],
        "uri": "timer",
      },
      "id": "kamelet-1234",
    },
  },
]
`;
